fastcrypto/src/secp256r1/mod.rs

```diff
@@ -1,4 +1,452 @@
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        let value = Self::decode_base64(&s).map_err(|e| de::Error::custom(e.to_string()))?;
+        Ok(value)
+    }
+}
+
+impl AsRef<[u8]> for Secp256r1PrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| Ok(self.privkey.to_bytes().into()))
+            .expect("OnceCell invariant violated")
+    }
+}
+
+impl zeroize::Zeroize for Secp256r1PrivateKey {
+    fn zeroize(&mut self) {
+        self.bytes.take().zeroize();
+        // SigningKey from the p256 crate implements zeroize on drop, so we do not need to zeroize it here.
+    }
+}
+
+impl Drop for Secp256r1PrivateKey {
+    fn drop(&mut self) {
+        self.bytes.take().zeroize();
+    }
+}
+
+impl zeroize::ZeroizeOnDrop for Secp256r1PrivateKey {}
+
+impl Serialize for Secp256r1Signature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.as_ref().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Secp256r1Signature {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let data: Vec<u8> = Vec::deserialize(deserializer)?;
+        <Secp256r1Signature as Signature>::from_bytes(&data)
+            .map_err(|e| de::Error::custom(e.to_string()))
+    }
+}
+
+impl Signature for Secp256r1Signature {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        if bytes.len() != SIGNATURE_SIZE {
+            return Err(Error::new());
+        }
+
+        let sig = ExternalSignature::try_from(bytes).map_err(|_| Error::new())?;
+
+        Ok(Secp256r1Signature {
+            sig,
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+impl Authenticator for Secp256r1Signature {
+    type PubKey = Secp256r1PublicKey;
+    type PrivKey = Secp256r1PrivateKey;
+    const LENGTH: usize = SIGNATURE_SIZE;
+}
+
+impl AsRef<[u8]> for Secp256r1Signature {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| {
+                Ok(<[u8; SIGNATURE_SIZE]>::try_from(self.sig.to_bytes().as_slice()).unwrap())
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+
+impl std::hash::Hash for Secp256r1Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl PartialEq for Secp256r1Signature {
+    fn eq(&self, other: &Self) -> bool {
+        self.sig == other.sig
+    }
+}
+
+impl Eq for Secp256r1Signature {}
+
+impl Display for Secp256r1Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{}", Base64::encode(self.as_ref()))
+    }
+}
+
+impl Default for Secp256r1Signature {
```

Comment:
do we really use the default anywhere?

Reply:
Yea, it's a bit annoying because it's a very odd trait to impl for both signatures, but it is used in sui, and I'm not sure whether it's possible to change the Sui code to not use default, but it would be great. The same is also true for public keys.